plutus-tx-plugin/test/Budget/andCheap.plc.golden

```diff
@@ -0,0 +1,52 @@
+let
+  data Bool | Bool_match where
+    True : Bool
+    False : Bool
+in
+letrec
+  data (List :: * -> *) a | Nil_match where
+    Nil : List a
+    Cons : a -> List a -> List a
+in
+letrec
+  !go : List Bool -> Bool
+    = \(ds : List Bool) ->
+        Nil_match
+          {Bool}
+          ds
+          {all dead. Bool}
+          (/\dead -> True)
+          (\(x : Bool)
+            (xs : List Bool) ->
+             /\dead ->
+               let
+                 !r : Bool = go xs
+               in
+               Bool_match
+                 x
+                 {all dead. Bool}
+                 (/\dead -> r)
+                 (/\dead -> False)
+                 {all dead. dead})
+          {all dead. dead}
+in
+let
+  !eta : List Bool
+    = (let
+          a = List Bool
+        in
+        \(c : Bool -> a -> a)
+         (n : a) ->
+          c
+            False
+            (c
+               True
+               (c
+                  True
+                  (c
+                     True
+                     (c True (c True (c True (c True (c True (c True n))))))))))
+        (\(ds : Bool) (ds : List Bool) -> Cons {Bool} ds ds)
```

Comment:
This won't parse back correctly. Also it'd be great to make it simply `Cons {Bool}`. I wonder if we should have an eta-contraction pass specifically for builtins (it's not safe to eta-contract things in the general case). We're wasting quite a bit here because of those lambdas.

Reply:
Ideally we'd inline it also. Another case of type-lets getting in the way!

> I wonder if we should have an eta-contraction pass specifically for builtins

There aren't any builtins here?

I think we could do this kind of eta-contraction fairly easily using our arity analysis, which currently implies that a value won't have any effects until it receives all its arguments. So in this instance we could deduce that `Cons {Bool}` is pure, which is what we need. Indeed I guess this is a generic enhancement to our purity analysis if we have arity information available.

Reply:
> Ideally we'd inline it also. 

Can it be less efficient? I think in theory the current alignment allows us to partially compute `Cons {Bool}` only once (where by "in theory" I mean "if we perform eta-contraction")? And it would be less efficient in terms of program size.

> There aren't any builtins here?

I confused `Cons` and `MkCons` :facepalm: 

> I think we could do this kind of eta-contraction fairly easily using our arity analysis, which currently implies that a value won't have any effects until it receives all its arguments. So in this instance we could deduce that `Cons {Bool}` is pure, which is what we need. Indeed I guess this is a generic enhancement to our purity analysis if we have arity information available.

That does sound very sensible.

Reply:
> This won't parse back correctly.

I don't think readable PIRs are supposed to be parseable, are they?

> eta-contraction pass

We have a ticket: https://input-output.atlassian.net/browse/PLT-1204

Reply:
> I don't think readable PIRs are supposed to be parseable, are they?

Right, we're not parsing readable syntax. I guess I should've said that the printed program doesn't convey the meaning of the original one.